c++/src/kj/io.h

```diff
@@ -85,6 +85,14 @@ class OutputStream {
   // Equivalent to write()ing each byte array in sequence, which is what the default implementation
   // does.  Override if you can do something better, e.g. use writev() to do the write in a single
   // syscall.
+
+#ifdef _WIN32
+  virtual void requireIsInBinaryMode();
```

Comment:
Hmm, I'm not so excited about polluting the streams APIs with this.

But I realize after some thought that we probably can't enforce binary mode on _all_ streams since we use `FdOutputStream` to write text to the console in some places.

Ideas:
* Only check in `writeMessageToFd()`. If people manually set up an `FdOutputStream` and use `writeMessage()`, then it's on them to ensure binary mode.
* The `write()` method could, in debug mode, check if the buffer being written contains NUL ('\0') bytes. If so, that's a strong indication that it's binary. Then we could warn/throw if the fd is in text mode. Almost all serialized capnp will contain some zeros so this should work well for detecting the problem in the Cap'n Proto use case. You could even limit it to checking only the first 8 bytes and you'd probably catch all capnp messages.
* We could always do the checking in `FdOutputStream`'s constructor, except when the output is stdout/stderr, which we'd assume is a legitimate place to write text. I'd be surprised if anyone's using `FdOutputStream` to write text to somewhere other than stdout/stderr... nevertheless this could still possibly break people.

Reply:
OK, I'll add the check in `writeMessageToFd()`, and additionally, in FdOutputStream::write, if the fourth byte is NUL, I check that the stream is in binary mode. Since it's unlikely that the message has more than 2^24 segments, this is likely to catch all non-packed Cap'n Proto messages being written. This is unlikely to catch packed messages, though. But this seems like a good trade-off.